Improvements to Charring Wood

StockiesLad opened the pull request:
I changed CharringWoodBlock#check to read the recipe manager so that custom fuels can be added for it. To make sure that things like slabs don't burn in a charcoal pit, I made a tag "charcoal_pile_valid_fuel" and only added logs to it. I tested with some custom log stacks in my modpack and it works pretty well.

This wasn't in the main PR as I wanted to personally test it in gameplay to ensure it doesn't just break everything.

StockiesLad commented:
A question I'd like your input on, is changing charring wood blocks to mimic the parent-block's model (and hitbox) whilst laying it's own texture. This, in conjunction with my previous changes, enables fully mod-compatible charcoal pits.

I'd like to get your approval first before working on that

jmb05 commented:
> A question I'd like your input on, is changing charring wood blocks to mimic the parent-block's model (and hitbox) whilst laying it's own texture. This, in conjunction with my previous changes, enables fully mod-compatible charcoal pits.

Sure, I see no problem in changing the Charring Wood Block in that way

StockiesLad commented:
> > A question I'd like your input on, is changing charring wood blocks to mimic the parent-block's model (and hitbox) whilst laying it's own texture. This, in conjunction with my previous changes, enables fully mod-compatible charcoal pits.
> 
> Sure, I see no problem in changing the Charring Wood Block in that way

Alright, perfect!

StockiesLad commented:
I think I managed to properly implement the mimic behavior. Before pushing this, I'd like to test it a bit more; on top of that, I have some embedded changes that I personally think makes the mod better that I can remove if you don't like them.

**Changes:**
1.  In my testing (and gameplay), I found that carbonize severely punishes big charcoal pits. And so I played around with a more logarithmic function to calculate the burn times. Which, I found to be much more balanced.
2. I added planks to `carbonize:charcoal_pile_valid_fuel` tag to allow people to fill up charcoal pits with more material.
3. Due to the modifications to how blockstates are transferred, I added charcoal variants of other wooden-based blocks, like fences & fencegates, etc (WIP).
5. Stacked Logs - These are technically what charcoal pits should be made of. It is crafted by putting logs in a chest-like pattern. They look something like this: ![146788](https://github.com/user-attachments/assets/c8405570-fa55-4ad5-83de-458dfb808e42)
6. Added 3 stages (blockstates) for charring wood: 'Igniting', 'Burning', 'Charring'. First stage copies the original model, second is the charring wood model and the third is the charcoal model.
7. Added small fire particles inside the block to make it make it more visually appealing. This and the previous change make it easier to observe what is actually happening.

jmb05 commented:
> 1. In my testing (and gameplay), I found that carbonize severely punishes big charcoal pits. And so I played around with a more logarithmic function to calculate the burn times. Which, I found to be much more balanced.

Sure I didn't do very much gameplay testing, so there is quite likely some balancing needed.



> 2. I added planks to `carbonize:charcoal_pile_valid_fuel` tag to allow people to fill up charcoal pits with more material.
> 
> 3. Due to the modifications to how blockstates are transferred, I added charcoal variants of other wooden-based blocks, like fences & fencegates, etc (WIP).

I have no problem with these additions.



> 4. Stacked Logs - These are technically what charcoal pits should be made of. It is crafted by putting logs in a chest-like pattern. They look something like this:

I'm not sure about this new block and if it is really needed... It seems to me that this would be a change that increases the throughput of charcoal pits (maybe intentionally), but i'd rather add as few blocks as possible to stay as close to vanilla as possible. I guess changing the throughput could be done by changing the time it takes. If we add this I would also change the model, I think it would be better to be a normal full block that looks like stacked logs (because the whole point of charcoal pits is to reduce oxygen of the fire and I think gaps between logs would be "unrealistic").



> 5. Added 3 stages (blockstates) for charring wood: 'Igniting', 'Burning', 'Charring'. First stage copies the original model, second is the charring wood model and the third is the charcoal model.

Is there any benefit with adding these stages?

jmb05 commented:
I also have no problem with more particles

StockiesLad commented:
> > 4. Stacked Logs - These are technically what charcoal pits should be made of. It is crafted by putting logs in a chest-like pattern. They look something like this:
> 
> I'm not sure about this new block and if it is really needed... It seems to me that this would be a change that increases the throughput of charcoal pits (maybe intentionally), but i'd rather add as few blocks as possible to stay as close to vanilla as possible. I guess changing the throughput could be done by changing the time it takes. If we add this I would also change the model, I think it would be better to be a normal full block that looks like stacked logs (because the whole point of charcoal pits is to reduce oxygen of the fire and I think gaps between logs would be "unrealistic").

You do make a good point here. The reason for this is, in my play-testing, charcoal pits are actually slightly over powered in any modding environment. Since a tree farm is pretty easy to setup, one can systemically burn logs for **huge** amounts of charcoal. My plan for this block was to enable higher throughput, with a balanced manner. This way, one could simply make the wood stack more expensive as plainly limiting logs can make the charcoal pits less capable.

Regarding the model, this is up to you. The previous image was a copy paste from the mod "Primalcore" as I wasn't on my computer at the time. My custom model is a little more full but ultimately similar. There are 3 reasons for this. 
- The first is to provide a default implementation of a non-cubic "mimic model" that one can follow should anyone want to add some custom recipes for charcoal pits. 
- Secondly, is making it shaped like that enables a more generic model as one can just think of it as a bunch of sticks, instead of making a stack for every wood type. This seems to line up with the minimalist nature of this mod.
- Lastly, is a little space is needed as the main concept (to my knowledge) is lowering the intake of oxygen by enclosing the pit, rather than packing everything as tightly as possible. But regardless, I'll send a picture of the model and I'll change it to how you like if you do end up wanting this block.

> 
> > 5. Added 3 stages (blockstates) for charring wood: 'Igniting', 'Burning', 'Charring'. First stage copies the original model, second is the charring wood model and the third is the charcoal model.
> 
> Is there any benefit with adding these stages?

I originally wanted to avoid adding blockstates as they are plain inefficient at doing their job. However, I noticed that the models will bug out a lot and refuse to update to the 'mimic texture' no matter what I did. This would cause the purple and black (missingno) texture to appear sometimes. Blockstates seemed to be the only thing that would actually update the model properly. And I figured if there are going to be stages, I would atleast make it worth while by using them to visually demonstrate the progress. In my testing it *did* successfully make charcoal pits a little less frustrating as sometimes you'd be wondering about the progress for these bigger charcoal pits. 

On top of just visuals, it allowed me to fix previous issues. Like how if a charcoal pit is already lit, one can switch out the surrounding blocks with whatever they want (like buttons). I mitigated this by altering the lightSide method to check if it's still 'igniting' - if it wasn't, it'd randomly (bound 100) replace the exposed charring wood with a fire block. This state check is important because if it weren't there, the charring wood would immediately attempt to start burning itself. Trying to use burning time checks *sometimes* worked here but produced inconsistent results like the models. I think that it has something to do with the fact the burntimes are not 100% synced.

P.S: I rewrote the PR description. It gives an overview of everything if you are interested. I also am going to push the commit soon after reviewing the code so you can review it directly.

StockiesLad commented:
Here is the model for the wood stack. Honestly I have a lot of other models that I was messing around with but I think this one is the best.
![image](https://github.com/user-attachments/assets/4155e4e7-7764-4bbd-bcdd-3e33aa56bc54)

StockiesLad commented:
The glass isn't a valid wall here I just replaced it quickly after igniting. Also I forgot to mention, the charcoal pit will play the fire charge sound when properly ignited as a auditory cue.

Igniting:
![image](https://github.com/user-attachments/assets/285bb5ef-c6c2-4613-b815-827990877c36)

Burning:
![image](https://github.com/user-attachments/assets/bd40f214-6e4e-4688-8d76-e82226d29ac8)

Charring:
![image](https://github.com/user-attachments/assets/aa490faf-8d7f-46c0-9b2f-de23798d1cd8)

StockiesLad commented:
Charcoal pits will work with different blocks & blockstates in the same pit as demonstrated here.

![image](https://github.com/user-attachments/assets/4071a7b2-0893-4778-826e-09812b6c6b59)
![image](https://github.com/user-attachments/assets/c260521a-86ce-450b-b064-63eb9dadcf87)
![image](https://github.com/user-attachments/assets/7d9e3c3f-f076-464a-a71c-88ece87aa1c8)
![image](https://github.com/user-attachments/assets/1b972136-0e21-4611-aa1e-97c6edc400bc)

StockiesLad commented:
Here it is! I did my best to try and split everything up into commits after reviewing the code myself.

jmb05 commented:
There seems to be an issue with falling charcoal stairs. When they land they drop as item basically duplicating themselves.

StockiesLad commented:
> There seems to be an issue with falling charcoal stairs. When they land they drop as item basically duplicating themselves.

Hmm weird. I don't think I've have that personally occur on my world. But, I'll look into it (I probably screwed the call order in one of the overrides). Thanks for letting me know!

StockiesLad commented:
Should be fixed now.

StockiesLad commented:
Alright I think I fixed the last point of potential crashes. I also tried to document things a little more to make everything easier to understand.

If there are things that you want me to change/explain, let me know.

P.S. I'm interested in helping maintain/update/develop this mod further if it's something you're on board with as well.
